enhancements/installer/vsphere-ipi-multiple-datacenter-clusters.md

```diff
@@ -0,0 +1,1169 @@
+  // +kubebuilder:validation:Pattern=^[a-z0-9]([-a-z0-9]*[a-z0-9])?$
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=63
+  // +kubebuilder:validation:Required
+  SecretNamespace string `json:"secretNamespace"`
+
+  // IP Family enables the ability to support IPv4 or IPv6
+  // Supported values are:
+  // ipv4 - IPv4 addresses only (Default)
+  // ipv6 - IPv6 addresses only
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxItems=2
+  IPFamilyPriority []string `json:"ipFamily,omitempty"`
+}
+
+// VSpherePlatformPlacementConstraint is the context information for VM placements within a failure domain
+type VSpherePlatformPlacementConstraint struct {
+  // resourcePool is the absolute path of the resource pool where virtual machines will be
+  // created. The absolute path is of the form /<datacenter>/host/<cluster>/Resources/<resourcepool>.
+  // TODO: maximum length?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxLength=1024
+  ResourcePool string `json:"resourcePool,omitempty"`
+
+  // folder is the name or inventory path of the folder in which the
+  // virtual machine is created/located.
+  // TODO: maximum length?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MaxLength=1024
+  Folder string `json:"folder,omitempty"`
+}
+
+// VSpherePlatformDeploymentZoneSpec holds the association between a
+// vCenter, failure domain and the virtual machine placementConstraints
+type VSpherePlatformDeploymentZoneSpec struct {
+
+  // TODO: name is the name of the VSpherePlatformDeploymentZoneSpec
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  // +kubebuilder:validation:Required
+  Name string `json:"name"`
+
+  // server is the fully-qualified domain name or the IP address of the vCenter server.
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MaxLength=255
+  Server string `json:"server,omitempty"`
+
+  // failureDomain is the name of the VSphereFailureDomain used for this VSphereDeploymentZone
+  // TODO: why would this be omitempty?
+  // TODO: What is the point of deploymentzonespec without a failure domain?
+  // +kubebuilder:validation:Optional
+  // +kubebuilder:validation:MinLength=1
+  // +kubebuilder:validation:MaxLength=256
+  FailureDomain string `json:"failureDomain,omitempty"`
+
+  // ControlPlane determines if this failure domain is suitable for use by control plane machines.
+  // TODO: we might use this
+  // TODO: it seems like this should be non-optional
+  // TODO: or default to false?
+  // +kubebuilder:validation:Optional
+  ControlPlane *bool `json:"controlPlane,omitempty"`
+
+  // PlacementConstraint encapsulates the placement constraints
+  // used within this deployment zone.
+  // +kubebuilder:validation:Required
+  PlacementConstraint VSpherePlatformPlacementConstraint `json:"placementConstraint"`
+}
+
+// TODO: below...
+
+type VSpherePlatformNodeNetworking struct {
```

Comment:
Rather than have all these options duplicate the internal/external names, would something like below be better:
```
networking:
  internal:
    networkSubnetCIDR: ...
    vmNetworkName: ...
    excludeNetworkSubnetCIDR: ... 
  external:
    networkSubnetCIDR: ...
    vmNetworkName: ...
    excludeNetworkSubnetCIDR: ... 
```
I'm not a huge fan of the names here btw, I think we can do better. Eg they all relate to VM IPs, but only one of them has VM in it, perhaps we can drop the VM

Reply:
fixed